module/state_synchronization/indexer/persister.go

```diff
@@ -0,0 +1,207 @@
+	}
+
+	p.LastPersistedSealedExecutionResult = p.executionResult
+
+	duration := time.Since(start)
+	log.Debug().
+		Dur("duration_ms", duration).
+		Msg("successfully prepared execution data for persistence")
+
+	return nil
+}
+
+// addEventsToBatch adds events from in-memory storage to the batch.
+// If there are no events, this is a no-op.
+func (p *Persister) addEventsToBatch(batch storage.Batch) error {
+	if eventsList := p.inMemoryEvents.Data(); len(eventsList) > 0 {
+		if err := p.events.BatchStore(p.executionResult.BlockID, []flow.EventsList{eventsList}, batch); err != nil {
+			return fmt.Errorf("could not add events to batch: %w", err)
+		}
+		p.log.Debug().Int("event_count", len(eventsList)).Msg("added events to batch")
+	}
+
+	return nil
+}
+
+// addResultsToBatch adds transaction results from in-memory storage to the batch.
+// If there are no results, this is a no-op.
+func (p *Persister) addResultsToBatch(batch storage.Batch) error {
+	if results := p.inMemoryResults.Data(); len(results) > 0 {
+		if err := p.results.BatchStore(p.executionResult.BlockID, results, batch); err != nil {
+			return fmt.Errorf("could not add transaction results to batch: %w", err)
+		}
+		p.log.Debug().Int("result_count", len(results)).Msg("added transaction results to batch")
+	}
+
+	return nil
+}
+
+// persistRegisters persists registers from in-memory to permanent storage.
+// Registers must be stored for every height, even if it's an empty set.
+func (p *Persister) persistRegisters() error {
+	registerData := p.inMemoryRegisters.Data()
+
+	// Always store registers for every height to maintain height continuity
+	if err := p.registers.Store(registerData, p.header.Height); err != nil {
+		return fmt.Errorf("could not persist registers: %w", err)
+	}
+
+	p.log.Debug().
+		Int("register_count", len(registerData)).
+		Msg("persisted registers")
+
+	return nil
+}
+
+// persistCollections persists collections from in-memory to permanent storage.
+// If there are no collections, this is a no-op.
+func (p *Persister) persistCollections(batch storage.Batch) error {
+	if collections := p.inMemoryCollections.LightCollections(); len(collections) > 0 {
+		if err := p.collections.BatchStoreLightAndIndexByTransaction(collections, batch); err != nil {
+			return fmt.Errorf("could not add collections to batch: %w", err)
+		}
+		p.log.Debug().Int("collections_count", len(collections)).Msg("added collections to batch")
+	}
+
+	return nil
+}
+
+// persistTransactions persists transactions from in-memory to permanent storage.
+// If there are no transactions, this is a no-op.
+func (p *Persister) persistTransactions(batch storage.Batch) error {
+	if transactions := p.inMemoryTransactions.Data(); len(transactions) > 0 {
+		if err := p.transactions.BatchStore(transactions, batch); err != nil {
+			return fmt.Errorf("could not add transactions to batch: %w", err)
+		}
+		p.log.Debug().Int("transactions_count", len(transactions)).Msg("added transactions to batch")
```

Comment:
rather than having separate log lines for each data type, how about adding a single debug in the main function that includes all of these counts.